source/validate_cfg.cpp

```diff
@@ -24,32 +24,303 @@
 // TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 // MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
 
+#include "validate.h"
 #include "validate_passes.h"
 
+#include <algorithm>
+#include <cassert>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+using std::find;
+using std::get;
+using std::make_pair;
+using std::numeric_limits;
+using std::pair;
+using std::transform;
+using std::unordered_map;
+using std::unordered_set;
+using std::vector;
+
+using libspirv::BasicBlock;
+
 namespace libspirv {
 
-// TODO(umar): Support for merge instructions
-// TODO(umar): Structured control flow checks
+namespace {
+
+using bb_ptr = BasicBlock*;
+using cbb_ptr = const BasicBlock*;
+using bb_iter = vector<BasicBlock*>::const_iterator;
+
+/// @brief Sorts the blocks in a CFG given the entry node
+///
+/// Returns a vector of basic block pointers in a Control Flow Graph(CFG) which
+/// are sorted in the order they were accessed in a post order traversal.
+///
+/// @param[in] entry the first block of a CFG
+/// @param[in] depth_hint a hint about the depth of the CFG
+///
+/// @return A vector of pointers in the order they were access in a post order
+/// traversal
+vector<const BasicBlock*> PostOrderSort(const BasicBlock& entry, size_t size) {
+  struct block_info {
+    cbb_ptr block;
+    bb_iter iter;
+  };
+
+  vector<cbb_ptr> out;
+  vector<block_info> staged;
+  unordered_set<uint32_t> processed;
+
+  staged.reserve(size);
+  staged.emplace_back(block_info{&entry, begin(entry.get_successors())});
```

Comment:
Why is `block_info{` necessary here?  Doesn't `emplace_back` do exactly that for you?